File: week02/pset2/02-readability/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

int count(int type, string text);
int Liau(int letters, int words, int sentences);

int main(void)
{
    string text = get_string("Text: ");
    int letters = count(1, text);
    int words = count(2, text);
    int sentences = count(3, text);
    int grade = Liau(letters, words, sentences);

    if (grade < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (grade >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %d\n", grade);
    }
}

int count(int type, string text)
{
    int counter = 0;
    for (int i = 0, length = strlen(text); i < length; i++)
    {
        if (type == 1 && isalpha(text[i])) // Count letters
        {
            counter++;
        }
        else if (type == 2 && isspace(text[i - 1])) // Count words
        {
            counter++;
        }
        else if (type == 3 &&
                 (text[i] == '.' || text[i] == '!' || text[i] == '?')) // Count sentences
        {
            counter++;
        }
    }
    if (type == 2)
    {
        counter++; // Plus 1 to last word
    }
    return counter;
}

int Liau(int letters, int words, int sentences)
{
    float L = ((float) letters / words) * 100;
    float S = ((float) sentences / words) * 100;
    float index = 0.0588 * L - 0.296 * S - 15.8;
    index = round(index);

    return index;
}

File: week06/pset6/sentimental-readability/readability.py
from cs50 import get_string
import math

def count_letters(text):
    """Counts the number of letters in the text."""
    count = 0
    for char in text:
        if char.isalpha():  # Check if the character is a letter
            count += 1
    return count

def count_words(text):
    """Counts the number of words in the text."""
    # Split the text by whitespace and filter out any empty strings
    words = text.split()
    return len(words)

def count_sentences(text):
    """Counts the number of sentences in the text."""
    count = 0
    for char in text:
        if char in ['.', '!', '?']:  # Check for sentence-ending punctuation
            count += 1
    return count

def calculate_grade(letters, words, sentences):
    """Calculates the Coleman-Liau index."""
    L = (letters / words) * 100  # Average letters per 100 words
    S = (sentences / words) * 100  # Average sentences per 100 words
    index = 0.0588 * L - 0.296 * S - 15.8
    return round(index)  # Round to nearest integer

def main():
    text = get_string("Text: ")  # Get user input
    letters = count_letters(text)  # Count letters
    words = count_words(text)  # Count words
    sentences = count_sentences(text)  # Count sentences

    # Calculate the grade level
    grade = calculate_grade(letters, words, sentences)

    # Output the grade level
    if grade < 1:
        print("Before Grade 1")
    elif grade >= 16:
        print("Grade 16+")
    else:
        print(f"Grade {grade}")

if __name__ == "__main__":
    main()
